### PR TITLE
[core] improve Color & Clamp performance

### PR DESCRIPTION
**Pull request by jonathanpeppers**
### Description of Change ###

When profiling startup of a HelloForms app on a Pixel 3 XL, I noticed:

    Total(ms) Self(ms)      Calls Method name
            3        1        572 Xamarin.Forms.Internals.NumericExtensions:Clamp (double,double,double)

This method is called multiple times for every `Color` created and
~142 are created on startup. This is why it shows up 572 times for an
app with a single `Label`.

I found there is a `Math.Clamp` implementation in corefx:

https://github.com/dotnet/runtime/blob/6662a0f2fd05298af1f9b1b020fa526595f336f7/src/libraries/System.Private.CoreLib/src/System/Math.cs#L224-L225

The only difference is this version can throw an exception, so we
could return the incoming value instead. That would be bad to change!

If I rework `NumericExtensions` to use corefx's implementation it is a
bit faster, I did a BenchmarkDotNet comparison running with Mono on
macOS:

     Method |      Mean |    Error |   StdDev |
    ------- |----------:|---------:|---------:|
     Clamp2 |  53.89 ns | 0.668 ns | 0.522 ns |
     Clamp1 |  61.84 ns | 1.270 ns | 2.289 ns |
     Color2 | 112.50 ns | 2.643 ns | 3.705 ns |
     Color1 | 129.03 ns | 2.603 ns | 4.760 ns |

Maybe every `Color` is ~13% faster?

Code for the benchmark is here:

https://github.com/jonathanpeppers/Benchmarks/blob/clamp/Benchmarks/Clamp.cs

Additionally, I reworked the list of default `Color` fields so that
they call a new private constructor that does less math and avoids
`Clamp` completely. We should still keep the original change, as it
would help any cases where the `Color.To*` methods would be used in
apps.

I seem to be able to see a small difference in a Release build running
on a Pixel 3 XL:

    Before:
    12-18 13:04:27.154  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +606ms
    12-18 13:04:30.851  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +589ms
    12-18 13:04:34.601  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +587ms
    12-18 13:04:38.352  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +575ms
    12-18 13:04:42.084  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +583ms
    12-18 13:04:45.802  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +573ms
    12-18 13:04:49.566  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +592ms
    12-18 13:04:53.284  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +583ms
    12-18 13:04:57.015  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +594ms
    12-18 13:05:00.715  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +581ms
    Average(ms): 586.3
    Std Err(ms): 3.05886689260364
    Std Dev(ms): 9.67298643990917

    After:
    12-18 13:35:38.745  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +572ms
    12-18 13:35:42.459  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +572ms
    12-18 13:35:46.209  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +581ms
    12-18 13:35:49.974  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +581ms
    12-18 13:35:53.724  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +574ms
    12-18 13:35:57.474  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +580ms
    12-18 13:36:01.207  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +573ms
    12-18 13:36:04.957  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +568ms
    12-18 13:36:08.707  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +566ms
    12-18 13:36:12.407  1490  1517 I ActivityTaskManager: Displayed com.xamarin.forms.helloforms/crc6450e568c951913723.MainActivity: +565ms
    Average(ms): 573.2
    Std Err(ms): 1.87853370714738
    Std Dev(ms): 5.94044517598546

This change seems low risk and would help all platforms.

### Issues Resolved ### 

None

### API Changes ###

None

### Platforms Affected ### 

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

You could test the difference on Android with my changes here: https://github.com/xamarin/Xamarin.Forms/pull/8867

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Core/NumericExtensions.cs [hunk 1]
```diff
@@ -1,20 +1,48 @@
...
 	public static class NumericExtensions
 	{
-
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
 		public static double Clamp(this double self, double min, double max)
 		{
-			return Math.Min(max, Math.Max(self, min));
+			if (max < min)
+			{
+				return self;
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/NumericExtensions.cs]: previous implementation used to return `max` in this case

**Comment by jonathanpeppers** [comment]
Yeah it's called 4 times for rgba colors: https://github.com/xamarin/Xamarin.Forms/blob/9d019b90ea76e7d85456523c1a848109c1d87cb3/Xamarin.Forms.Core/Color.cs#L101-L104

Then there is a list of ~142 colors: https://github.com/xamarin/Xamarin.Forms/blob/9d019b90ea76e7d85456523c1a848109c1d87cb3/Xamarin.Forms.Core/Color.cs#L421-L564

I have another idea that does the math for these at build time: https://github.com/xamarin/Xamarin.Forms/compare/master...jonathanpeppers:colorgen

Still working on that one ☝️

**Comment by StephaneDelcroix** [comment]
@jonathanpeppers that looks cool, but a lot of work for very little benefits.

we could:
- have a private ctor that bypass the Clamp,
- or precompute each value and directly assign the components (precompute the HSL components too)

**Comment by jonathanpeppers** [comment]
@StephaneDelcroix instead of generating the `cs` file like this: https://github.com/xamarin/Xamarin.Forms/compare/master...jonathanpeppers:colorgen

Do these color change often? As I could just run this "script" once, and we just commit the list of colors with precomputed `float` values.

**Comment by StephaneDelcroix** [comment]
they never change. committing the result (with the ld values as comment) would be good

**Comment by jonathanpeppers** [comment]
That idea didn't actually pan out: https://github.com/xamarin/Xamarin.Forms/compare/4.4.0...jonathanpeppers:precompute-color-floats

My guess is the JIT time + actual larger data is slower than the math that happens here.

**Comment by StephaneDelcroix** [comment]
but we could still bypass the Clamp for predefined colors

**Comment by jonathanpeppers** [comment]
> but we could still bypass the Clamp for predefined colors

Yeah, let me try defining the three floats for RGB and still do the math for HSL.

File: Xamarin.Forms.Core/Color.cs [hunk 2]
```diff
@@ -114,6 +114,26 @@ public Color(double r, double g, double b, double a) : this(r, g, b, a, Mode.Rgb
...
+
+		Color(byte r, byte g, byte b, byte a)
+		{
+			_mode = Mode.Rgb;
+			_r = r / 255f;
+			_g = g / 255f;
+			_b = b / 255f;
+			_a = a / 255f;
+			ConvertToHsl(_r, _g, _b, _mode, out _hue, out _saturation, out _luminosity);
+		}
```
Comment (jonathanpeppers) [inline Xamarin.Forms.Core/Color.cs]: @StephaneDelcroix these new constructors here performed better than precomputing the `float` values: https://github.com/jonathanpeppers/Xamarin.Forms/commit/607b977cc86858e4fccd68575febdb423fa83102

It seems like the huge list of `float` always made things worse.

**Comment by ahmedalejo** [comment]
Hi @jonathanpeppers could you benchmark the following.
Most of the existent colors aren´t used in a common app´s life-cycle (at least not at startup)

```csharp
public partial struct Color
{
	private static readonly Color aliceBlue;
	public static Color AliceBlue => aliceBlue ??= new Color(240, 248, 255);
...
```

**Comment by jonathanpeppers** [comment]
@ahmedalejo can you open a PR? (or link to a diff) I can time the difference.

The main issue is the API change, going from a field to a property can break existing code.

**Comment by StephaneDelcroix** [comment]
> going from a field to a property can break existing code.

it WILL break ABI. We can't do that

**Comment by ahmedalejo** [comment]
Hi @StephaneDelcroix 
- i get your point. But "ABI" really meant "API" and not "Android ABI" right?

Hi @jonathanpeppers here a pull request
- https://github.com/xamarin/Xamarin.Forms/pull/9109

**Comment by ahmedalejo** [comment]
And a json representation of the colors that helped with code gen in linqpad:
```csharp
void Main()
{
    var colorsJson= Newtonsoft.Json.Linq.JObject.Parse(json);

    foreach (var p in colorsJson.Properties())
        Console.WriteLine($"\t\tpublic static Color {CamelCase(p.Name)};");

    Console.WriteLine();
    foreach (var p in colorsJson.Properties())
        Console.WriteLine($"        public static Color {p.Name} => {CamelCase(p.Name)} ??= {p.Value};");

    string CamelCase(string value) 
        => value[0].ToString().ToLower() + value[1..];
}
```
```json
{
    "AliceBlue" : "new Color(240, 248, 255)",
    "AntiqueWhite" : "new Color(250, 235, 215)",
    "Aqua" : "new Color(0, 255, 255)",
    "Aquamarine" : "new Color(127, 255, 212)",
    "Azure" : "new Color(240, 255, 255)",
    "Beige" : "new Color(245, 245, 220)",
    "Bisque" : "new Color(255, 228, 196)",
    "Black" : "new Color(0, 0, 0)",
    "BlanchedAlmond" : "new Color(255, 235, 205)",
    "Blue" : "new Color(0, 0, 255)",
    "BlueViolet" : "new Color(138, 43, 226)",
    "Brown" : "new Color(165, 42, 42)",
    "BurlyWood" : "new Color(222, 184, 135)",
    "CadetBlue" : "new Color(95, 158, 160)",
    "Chartreuse" : "new Color(127, 255, 0)",
    "Chocolate" : "new Color(210, 105, 30)",
    "Coral" : "new Color(255, 127, 80)",
    "CornflowerBlue" : "new Color(100, 149, 237)",
    "Cornsilk" : "new Color(255, 248, 220)",
    "Crimson" : "new Color(220, 20, 60)",
    "Cyan" : "new Color(0, 255, 255)",
    "DarkBlue" : "new Color(0, 0, 139)",
    "DarkCyan" : "new Color(0, 139, 139)",
    "DarkGoldenrod" : "new Color(184, 134, 11)",
    "DarkGray" : "new Color(169, 169, 169)",
    "DarkGreen" : "new Color(0, 100, 0)",
    "DarkKhaki" : "new Color(189, 183, 107)",
    "DarkMagenta" : "new Color(139, 0, 139)",
    "DarkOliveGreen" : "new Color(85, 107, 47)",
    "DarkOrange" : "new Color(255, 140, 0)",
    "DarkOrchid" : "new Color(153, 50, 204)",
    "DarkRed" : "new Color(139, 0, 0)",
    "DarkSalmon" : "new Color(233, 150, 122)",
    "DarkSeaGreen" : "new Color(143, 188, 143)",
    "DarkSlateBlue" : "new Color(72, 61, 139)",
    "DarkSlateGray" : "new Color(47, 79, 79)",
    "DarkTurquoise" : "new Color(0, 206, 209)",
    "DarkViolet" : "new Color(148, 0, 211)",
    "DeepPink" : "new Color(255, 20, 147)",
    "DeepSkyBlue" : "new Color(0, 191, 255)",
    "DimGray" : "new Color(105, 105, 105)",
    "DodgerBlue" : "new Color(30, 144, 255)",
    "Firebrick" : "new Color(178, 34, 34)",
    "FloralWhite" : "new Color(255, 250, 240)",
    "ForestGreen" : "new Color(34, 139, 34)",
    "Fuchsia" : "new Color(255, 0, 255)",
    "Fuschia" : "new Color(255, 0, 255)",
    "Gainsboro" : "new Color(220, 220, 220)",
    "GhostWhite" : "new Color(248, 248, 255)",
    "Gold" : "new Color(255, 215, 0)",
    "Goldenrod" : "new Color(218, 165, 32)",
    "Gray" : "new Color(128, 128, 128)",
    "Green" : "new Color(0, 128, 0)",
    "GreenYellow" : "new Color(173, 255, 47)",
    "Honeydew" : "new Color(240, 255, 240)",
    "HotPink" : "new Color(255, 105, 180)",
    "IndianRed" : "new Color(205, 92, 92)",
    "Indigo" : "new Color(75, 0, 130)",
    "Ivory" : "new Color(255, 255, 240)",
    "Khaki" : "new Color(240, 230, 140)",
    "Lavender" : "new Color(230, 230, 250)",
    "LavenderBlush" : "new Color(255, 240, 245)",
    "LawnGreen" : "new Color(124, 252, 0)",
    "LemonChiffon" : "new Color(255, 250, 205)",
    "LightBlue" : "new Color(173, 216, 230)",
    "LightCoral" : "new Color(240, 128, 128)",
    "LightCyan" : "new Color(224, 255, 255)",
    "LightGoldenrodYellow" : "new Color(250, 250, 210)",
    "LightGray" : "new Color(211, 211, 211)",
    "LightGreen" : "new Color(144, 238, 144)",
    "LightPink" : "new Color(255, 182, 193)",
    "LightSalmon" : "new Color(255, 160, 122)",
    "LightSeaGreen" : "new Color(32, 178, 170)",
    "LightSkyBlue" : "new Color(135, 206, 250)",
    "LightSlateGray" : "new Color(119, 136, 153)",
    "LightSteelBlue" : "new Color(176, 196, 222)",
    "LightYellow" : "new Color(255, 255, 224)",
    "Lime" : "new Color(0, 255, 0)",
    "LimeGreen" : "new Color(50, 205, 50)",
    "Linen" : "new Color(250, 240, 230)",
    "Magenta" : "new Color(255, 0, 255)",
    "Maroon" : "new Color(128, 0, 0)",
    "MediumAquamarine" : "new Color(102, 205, 170)",
    "MediumBlue" : "new Color(0, 0, 205)",
    "MediumOrchid" : "new Color(186, 85, 211)",
    "MediumPurple" : "new Color(147, 112, 219)",
    "MediumSeaGreen" : "new Color(60, 179, 113)",
    "MediumSlateBlue" : "new Color(123, 104, 238)",
    "MediumSpringGreen" : "new Color(0, 250, 154)",
    "MediumTurquoise" : "new Color(72, 209, 204)",
    "MediumVioletRed" : "new Color(199, 21, 133)",
    "MidnightBlue" : "new Color(25, 25, 112)",
    "MintCream" : "new Color(245, 255, 250)",
    "MistyRose" : "new Color(255, 228, 225)",
    "Moccasin" : "new Color(255, 228, 181)",
    "NavajoWhite" : "new Color(255, 222, 173)",
    "Navy" : "new Color(0, 0, 128)",
    "OldLace" : "new Color(253, 245, 230)",
    "Olive" : "new Color(128, 128, 0)",
    "OliveDrab" : "new Color(107, 142, 35)",
    "Orange" : "new Color(255, 165, 0)",
    "OrangeRed" : "new Color(255, 69, 0)",
    "Orchid" : "new Color(218, 112, 214)",
    "PaleGoldenrod" : "new Color(238, 232, 170)",
    "PaleGreen" : "new Color(152, 251, 152)",
    "PaleTurquoise" : "new Color(175, 238, 238)",
    "PaleVioletRed" : "new Color(219, 112, 147)",
    "PapayaWhip" : "new Color(255, 239, 213)",
    "PeachPuff" : "new Color(255, 218, 185)",
    "Peru" : "new Color(205, 133, 63)",
    "Pink" : "new Color(255, 192, 203)",
    "Plum" : "new Color(221, 160, 221)",
    "PowderBlue" : "new Color(176, 224, 230)",
    "Purple" : "new Color(128, 0, 128)",
    "Red" : "new Color(255, 0, 0)",
    "RosyBrown" : "new Color(188, 143, 143)",
    "RoyalBlue" : "new Color(65, 105, 225)",
    "SaddleBrown" : "new Color(139, 69, 19)",
    "Salmon" : "new Color(250, 128, 114)",
    "SandyBrown" : "new Color(244, 164, 96)",
    "SeaGreen" : "new Color(46, 139, 87)",
    "SeaShell" : "new Color(255, 245, 238)",
    "Sienna" : "new Color(160, 82, 45)",
    "Silver" : "new Color(192, 192, 192)",
    "SkyBlue" : "new Color(135, 206, 235)",
    "SlateBlue" : "new Color(106, 90, 205)",
    "SlateGray" : "new Color(112, 128, 144)",
    "Snow" : "new Color(255, 250, 250)",
    "SpringGreen" : "new Color(0, 255, 127)",
    "SteelBlue" : "new Color(70, 130, 180)",
    "Tan" : "new Color(210, 180, 140)",
    "Teal" : "new Color(0, 128, 128)",
    "Thistle" : "new Color(216, 191, 216)",
    "Tomato" : "new Color(255, 99, 71)",
    "Transparent" : "new Color(255, 255, 255, 0)",
    "Turquoise" : "new Color(64, 224, 208)",
    "Violet" : "new Color(238, 130, 238)",
    "Wheat" : "new Color(245, 222, 179)",
    "White" : "new Color(255, 255, 255)",
    "WhiteSmoke" : "new Color(245, 245, 245)",
    "Yellow" : "new Color(255, 255, 0)",
    "YellowGreen" : "new Color(154, 205, 50)"
}
```

**Pull request merged**